src/shared/GuaranteeDelivery.js

```diff
@@ -0,0 +1,173 @@
+export const INTERVALS_DELAY = 100; // 100ms.
+
+const ACK_FIELD = '__gack__';
+const FROM_FIELD = '__from__';
+const TO_ACK_FIELD = '__to_ack__';
+const RETRY_FIELD = '__gretry__';
+const TRANSACTION_FIELD = 'transaction';
+export const DEADLINE_EXCEEDED = 'DEADLINE_EXCEEDED';
+
+// Class for sending and receiving messages that we validate they were eventually received. If not we raise an error after MAX_DELAY.
+export class GuaranteeDeliveryManager {
+	constructor(userId, maxDelay, retryDelay, intervalsDelay) {
+    this.maxDelay = maxDelay;
+    this.retryDelay = retryDelay;
+    this.intervalsDelay = intervalsDelay;
+
+    // Map from transaction to timestamp that message was sent and promise resolving or success
+    // or rejecting on failure.
+    // @const {!Map<string, {delivery: !Promise, sent: number, sendMessage: function, message: Object!}>}
+    this.pending = new Map();
+
+    // Map from transaction to timestamp guarantee message was received.
+    this.accepted = new Map();
+
+    // Process pending messages status.
+    this.intervalIdentifier = setInterval(() => this.iterate_(), this.intervalsDelay);
+
+    if (!userId) {
+      throw new Error(`Unique userId has to be set for GuaranteeDeliveryManager, got [${userId}].`);
+    }
+    this.userId = userId;
+	}
+
+  // Iterate over exising pending messages and retries them if needed, when max delay
+  // passes, delete them and reject.
+  iterate_() {
+    const now = Number(new Date());
+    for (let [transaction, state] of this.pending) {
+      const interval = now - state.sent;
+      if (interval > this.maxDelay) {
+        console.log('Did not get all ack as required', state.ack);
+        // Failed sending message on timeout. Delete and reject.
+        this.pending.delete(transaction);
+        state.delivery.reject({reason: DEADLINE_EXCEEDED, message: state.message});  // Note that if reject is computation heavy action, our timings might be not accurate.
+      } else if (now - state.sent >= (state.message[RETRY_FIELD] + 1) * this.retryDelay) {
+        // We should retry.
+        state.message[RETRY_FIELD]++;
+        state.sendMessage(state.message);
+      }
+    }
+
+    // Garbage collect accepted messages.
+    for (let [transaction, accepted] of this.accepted) {
+      const interval = now - accepted;
+      if (interval > 10 * this.maxDelay) {
+        // We don't expect messages to arrive after 10 * MAX_DELAY time, it is safe to delete
+        // them now.
+        this.accepted.delete(transaction);
+      }
+    }
+  }
+
+  // Send message that guarantee to arrive within MAX_DELAY or notify on timeout.
+  // @param {!Object} message
+  // @param {!Array<string>} toAck List of userIds that are expected to ack. If empty will expect any non-self user to ack.
+  // @param {function(!Object)} sendMessage
+  // @return {!Promise}  <====  Resolve for success, reject for failure.
+  sendGuaranteeMessage(message, toAck, sendMessage) {
```

Comment:
I'd use shorter names: `send`, `accept` and `checkAck`.
The words 'Guarantee' and 'Message' are implicit for users of this class